shared/js/dialer/contacts.js

```diff
@@ -11,6 +11,153 @@ var Contacts = {
     '/shared/js/fb/fb_reader_utils.js'
   ],
 
+  _calledNumberCache: [],
+
+  _CACHE_SIZE: 10,
+
+  init: function init() {
+    var self = this;
+    var mozContacts = window.navigator.mozContacts;
+
+    asyncStorage.getItem('contactsCache', function (value) {
+      if(value) {
+        self._calledNumberCache = JSON.parse(value);
+      }
+    });
+
+    if (mozContacts) {
+      LazyLoader.load(self._FB_FILES, function() {
+        mozContacts.oncontactchange = self._oncontactchange.bind(self);
```

Comment:
This is not enough to track changes properly. You also have to track the contact manager's revision number using ``mozContacts.getRevision()``. If the revision changed between two init() calls you have to invalidate the entire cache because the contacts will have changed when your ``oncontactchange`` is not registered (for example because the callscreen wasn't running).